numpy/random/_generator.pyx

```diff
@@ -4150,17 +4156,32 @@ cdef class Generator:
 
         i = 0
         totsize = np.PyArray_SIZE(val_arr)
-        with self.lock, nogil:
-            while i < totsize:
-                acc = 0.0
-                for j in range(k):
-                    val_data[i+j] = random_standard_gamma(&self._bitgen,
-                                                              alpha_data[j])
-                    acc = acc + val_data[i + j]
-                invacc = 1/acc
-                for j in range(k):
-                    val_data[i + j] = val_data[i + j] * invacc
-                i = i + k
+
+        # Small alpha case: Use stick-breaking approach with beta random variates
```

Comment:
PEP 8: Keep line lengths 79 characters or less (here and elsewhere).

Reply:
Fixed.

Reply:
@WarrenWeckesser : Thanks a lot for the extensive review. So far I have been able to address the following:
* Adding the suggested test
* Fixing the formatting issues.

As for documenting the rationale in the code change: Thanks a lot for your suggestions. I haven't found time to do a formal analysis giving the probability of generating gamma variates which sum to zero for a given vector of alpha parameters. That could potentially help to determine a new threshold for the switch between the two code branches. 

If I remember correctly from the past discussion the point of adding the new code branch is to ensure that the old code branch is never called when there is positive probability of returning NaN. I think this is why we ended up with one being the threshold in the first place, but I'd need to backtrack the argument again to be 100% sure.

Reply:
I'm fine with keeping the threshold at 1.  @bashtage, @rkern, any opinions?

Can you expand the comments in the code with a sentence or two about why there are two methods?  E.g. the normalized gamma variates method is faster than the "stick breaking" method, but it generates nans when all the gamma variates happen to be 0, and that can happen when all the alpha values are sufficiently small.

Reply:
Will do that, as soon as possible.

Reply:
> I'm fine with keeping the threshold at 1. @bashtage, @rkern, any opinions?

One argument for a smaller alpha is that there is no need to unnecessarily break the stream.  If someone uses alpha 0.5 then the will get a new stream now even though the old method was considered correct.

I think an ideal threshold would be the smallest alpha where (a) [essential] accuracy is high and (b) performance is good.

This said, I don't have much of a prior for values that Dirichlet users are actually using, so perhaps [1e-4,1) has little probability.

Reply:
I have added an inline docstring that provides the rationale for the switch to the stick-breaking approach. 

As for the choice of the threshold, the probability for a gamma RV to be smaller than the smallest postive floating precision number can be computed using the regularized incomplete gamma function

```python
import numpy as np
from scipy.special import gammainc
delta = np.finfo(np.float).tiny
delta
params = [1.0e-5, 1.0e-4, 1.0e-3, 1.0e-2, 1.0e-1, 1.]
probs = [gammainc(p, delta) for p in params]
probs
>>>[0.9929467992914456,
 0.9316650395588774,
 0.4927171386203597,
 0.0008432274068068666,
 1.8046532550200846e-31,
 2.2250738585072626e-308]
```

From the number one could argue that if alpha.max() >= 0.1 the probability of generating a single zero is still so tiny that it can be neglected (the probability of generating all zeros is still smaller or equal) and the threshold can be reduced to 0.1

Not sure if that is small enough though?

Reply:
Nice. I was working through similar reasoning, using `np.finfo(np.float).tiny` as the threshold for "numerically zero", and came to roughly the same conclusion.  With the revised threshold condition `max(alpha) > 0.1`, a conservative upper bound on the probability of generating all  "tiny" gamma variates is 1.8e-31, so I think a threshold of 0.1 should be safe.  For comparsion, here are some other values for `max(alpha)` and the associated upper bounds for the probabilty that all the gamma variates are "tiny"

```
   max(alpha)      p
   ----------  ---------
   0.05        4.3e-16
   0.1         1.8e-31
   0.2         3.2e-62
   0.25        1.3e-77
   0.5         1.7e-154
   1.0         2.2e-308
```
@bashtage, what do you think?  Does the threshold 0.1 sound reasonable?

Reply:
I have changed the threshold from 1. to 0.1 to 
i) Not unneccesarily 'break the stream' compared to the legacy method of generation 
ii) Use the method with a higher performance whenever the probability of failure is acceptable

From the previous discussion I understand that a probabilty of failure < 1.8e-31 is acceptable.

Reply:
Setting the threshold to 0.1 (so the estimate of the upper bound of the probability that the gamma variates method produces `nan` is 1.8e-31) seems fine, but to be honest, I don't have a rigorous or even heuristic argument for what we should consider acceptable.  How about this... On my computer, `dirichlet` with `alpha=[0.05, 0.5, 1.5]` generates approximately 7 million samples per second.  After running for a million years (single-threaded), I would generate only about 2.2e20 samples, so the probability that I would get all 0s in a sample is still negligible.

Don't forget to also update the unit test.

Reply:
I have updated the unit tests to respect the updated threshold.